### PR TITLE
✨ Report amp4ads-id in getErrorReportData() for inabox

### LINKED ISSUE #18004: Report amp4ads-id in error messages
`amp4ads-id` is introduced in #7730 . it can be used in error message to identify the ad creative.

Tip: the change will be in error.js#getErrorReportData

### PR DESCRIPTION
**Pull request by torch2424**
closes #18004
relates to #8051

This reports an `a4aid` in the data sent to servers for error collection. The entire string from the `content` attribute is sent in the request.

Also, this adds `a4aId` to the `AMP_MODE` if we have the correct a4a meta tag.

# Example

Modified `examples/inabox.host.html` to achieve these logs.

screen shot 2018-09-19 at 6 09 00 pm

### CODE REVIEW

**Comment by zhouyx** [comment]
@torch2424  @lannka Should we move the id to [here](https://github.com/ampproject/amphtml/blob/master/src/error.js#L278)?

**Comment by torch2424** [comment]
I think moving it there would be good! Is there any requirements on the object key on window? Is it alright if it just becomes `window.a4aId` ? cc @zhouyx

File: src/mode.js [hunk 1]
```diff
@@ -155,3 +158,24 @@ export function getRtvVersionForTesting(win, isLocalDev) {
 export function resetRtvVersionForTesting() {
   rtvVersion = '';
 }
+
+/**
+ * Function to get the amp4ads-identifier from the meta tag on the document
+ * @param {!Window} win
+ * @return {?string}
+ */
+function getA4AId_(win) {
```
Comment (lannka) [inline src/mode.js]: move the logic to amp-inabox.js

File: src/mode.js [hunk 2]
```diff
@@ -155,3 +158,24 @@ export function getRtvVersionForTesting(win, isLocalDev) {
...
+
+  if (!win.document || !win.document.querySelector) {
+    return null;
+  }
+
+  const a4aIdMetaTag = win.document
```
[the remark this answers is not in the document]
Comment (torch2424) [reply src/mode.js]: Done

File: src/mode.js (hunk 1 above)
Comment (torch2424) [reply src/mode.js]: Done

**Comment by torch2424** [comment]
cc @lannka @zhouyx This is good to go 😄 

Agreed to move this into `amp-inabox.js` with @lannka offline. And moved the function into utils, for testing, and to abstract out this logic from the main code.

File: src/error.js [hunk 3]
```diff
@@ -485,6 +489,11 @@ export function getErrorReportData(message, filename, line, col, error,
   }
   data['rt'] = runtime;
 
+  // Add our a4a id if we are inabox
+  if (runtime === 'inabox') {
+    data['a4aId'] = getMode().a4aId;
```
Comment (lannka) [inline src/error.js]: nit: adid for short

File: test/functional/inabox/test-utils.js [hunk 4]
```diff
@@ -52,10 +68,21 @@ describes.fakeWin('inabox-utils', {}, env => {
     setTimeout(() => timeDeferred.resolve(), 10);
     return timeDeferred.promise.then(() => {
       expect(dispatchEventStub).to.be.calledOnce;
-      expect(initCustomEventStub.withArgs('amp-ini-load', true, false, null))
+      expect(initCustomEventStub.withArgs('amp-ini-load'))
```
Comment (lannka) [inline test/functional/inabox/test-utils.js]: you can event do
```
expect(abc).to.be.calledWith('123');
```

File: test/functional/inabox/test-utils.js [hunk 5]
```diff
@@ -52,10 +68,21 @@ describes.fakeWin('inabox-utils', {}, env => {
...
+
+  it('Should not return an a4aId if no a4a meta tag in head', () => {
+    expect(getA4AId(env.win)).to.be.not.ok;
+  });
+
+
+  it('Should be able to get the a4aId if on the document', () => {
+    addA4AMetaTagToDocument();
+    expect(getA4AId(env.win)).to.be.ok;
+    removeA4AMetaTagToDocument();
```
Comment (lannka) [inline test/functional/inabox/test-utils.js]: you don't need to clean up, given every iframe that runs the test is new.

File: test/functional/inabox/test-utils.js [hunk 6]
```diff
@@ -36,8 +49,11 @@ describes.fakeWin('inabox-utils', {}, env => {
     dispatchEventStub = sandbox.stub();
     initCustomEventStub = sandbox.stub();
     env.win.parent = {postMessage: parentPostMessageStub};
-    env.win.document =
-      {createEvent: () => ({initCustomEvent: initCustomEventStub})};
+    env.win.CustomEvent = (type, eventInit) => {
```
Comment (lannka) [inline test/functional/inabox/test-utils.js]: why is this needed?

File: test/functional/inabox/test-utils.js [hunk 7]
```diff
@@ -16,14 +16,27 @@
...
+describes.realWin('inabox-utils', {}, env => {
   let getResourcesInRectStub;
   let dispatchEventStub;
   let parentPostMessageStub;
   let initCustomEventStub;
   let ampdoc;
+  let a4aIdMetaElement;
+
+  function addA4AMetaTagToDocument() {
+    a4aIdMetaElement = env.win.document.createElement('meta');
```
Comment (lannka) [inline test/functional/inabox/test-utils.js]: you can use: dom.js#createElementWithAttributes

File: src/error.js (hunk 3 above)
Comment (torch2424) [reply src/error.js]: Done.

**Comment by zhouyx** [comment]
So we've agreed to have the id as part of the mode? LGTM since it is only in inabox. 

LGTM Thank you @torch2424, the change is going to be really helpful!!

File: test/functional/inabox/test-utils.js (hunk 7 above)
Comment (torch2424) [reply test/functional/inabox/test-utils.js]: Done.

File: test/functional/inabox/test-utils.js (hunk 4 above)
Comment (torch2424) [reply test/functional/inabox/test-utils.js]: Done

File: test/functional/inabox/test-utils.js (hunk 5 above)
Comment (torch2424) [reply test/functional/inabox/test-utils.js]: Done.

File: test/functional/inabox/test-utils.js (hunk 6 above)
Comment (torch2424) [reply test/functional/inabox/test-utils.js]: This is to mock out [CustomEvent](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent) on the window. Which is the newer API of of creating events, relative to [createEvent](https://developer.mozilla.org/en-US/docs/Web/API/Document/createEvent).

Since this test now uses a `realWin` instead of a `fakeWin`, it no longer will use the fallback, thus we need to mock out the modern use case.

**Comment by torch2424** [comment]
@zhouyx 

Yes, we agreed mode is a good place, since it contains many different stateful information about the AMP document. However, made it so that it only checks to add this to mode in the inabox.js 😄 

@zhouyx , @lannka 

This is good to go for another review!

File: src/error.js [hunk 8]
```diff
@@ -489,6 +489,11 @@ export function getErrorReportData(message, filename, line, col, error,
   }
   data['rt'] = runtime;
 
+  // Add our a4a id if we are inabox
+  if (runtime === 'inabox') {
+    data['adid'] = getMode().a4aId;
```
Comment (erwinmombay) [inline src/error.js]: need to add this field to https://github.com/ampproject/amphtml/blob/master/src/mode.js#L32
fixing the single pass code to actually throw the error

Comment (erwinmombay) [reply src/error.js]: in saying that, `runtime` also really needs to be there. the reason why it doesnt cause an error is because we do a `getMode().runtime` conditional before reading a few lines above this

Comment (torch2424) [reply src/error.js]: Ahhh makes sense. Thank you so much for tracking this down, I'll go ahead and fix this now 👍

Comment (torch2424) [reply src/error.js]: Done.
